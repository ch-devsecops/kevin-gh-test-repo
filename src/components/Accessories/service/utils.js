import { useTranslation } from 'react-i18next';

import { flatten } from 'lodash';
import { isEmpty } from '../../../utils/object';
import {
  variant1,
  variant2,
  variant3,
  ALL_TAB,
  INTERIOR_TAB,
  EXTERIOR_TAB,
  PACKAGES_TAB,
  TRIM_DROPDOWN_KEY,
  TRANSMISSION_DROPDOWN_KEY,
  ACCESSORY_TYPE_DROPDOWN_KEY,
} from './constants';
import { useDarkFromRoute } from '../../../utils/sitecoreContext';

/**
 * Formatter used in fetching Accessories for Autos
 *
 * @param {array} Accessories - The array of Accessories from each tab.
 * @returns {object} - Compiles the Accessories arrays into an object containing the respective arrays
 */
export const formatAccessories = accessories => {
  const packageAccessories = {
    interior: [],
    exterior: [],
    packages: [],
  };

  accessories.forEach(accessory => {
    if (packageAccessories[accessory?.accessoryCategory?.key]) {
      packageAccessories[accessory?.accessoryCategory?.key]?.push(accessory);
    }
  });

  return packageAccessories;
};

/**
 * Formatter used in fetching Accessories for PSP or for any other Products with generic category
 *
 * @param {array} Accessories - The array of Accessories
 *  @returns {object} - Compiles the Accessories array into an object
 */
export const formatAccessoriesGenericCategory = accessories => {
  const packageAccessories = {};

  accessories.forEach(accessory => {
    if (!packageAccessories[accessory?.accessoryCategory?.key]) {
      packageAccessories[accessory?.accessoryCategory?.key] = [];
    }
    packageAccessories[accessory?.accessoryCategory?.key]?.push(accessory);
  });

  return packageAccessories;
};

/**
 * Stores a variety of props which change value depending on what site the user is on
 *
 * @param {variant} string - String containing the variant name
 *  @returns {object} - Returns an object containing the prop values based on the variant
 */
export const useConfiguration = variant => {
  const { t } = useTranslation();
  const isDark = useDarkFromRoute();
  let dropdownKeys = [TRIM_DROPDOWN_KEY];
  let showTabs = true;
  let showAccordion = true;
  let showSlider = false;
  let defaultTabs = [
    {
      key: INTERIOR_TAB,
      title: t('Pages.Models.Exploration.interiorLabel'),
    },
    {
      key: EXTERIOR_TAB,
      title: t('Pages.Models.Exploration.exteriorLabel'),
    },
    {
      key: PACKAGES_TAB,
      title: t('Pages.Models.Exploration.packagesLabel'),
    },
  ];
  let defaultTab = INTERIOR_TAB;
  let trimPathKey = 'defaultTransmission.detIdentifier';
  let trimTransmissionPathKey = '';
  let optionTrimValuePath = 'defaultTransmission.detIdentifier';
  let transmissionStateKey = 'submittedTrim';
  let accessoryFormatter = formatAccessories;
  const dropdownList = [
    {
      key: TRIM_DROPDOWN_KEY,
      label: t('Pages.Models.Exploration.trimLabel'),
      placeholder: t('Pages.Models.Exploration.trimLabel'),
    },
  ];
  let hasAsterisk = false;
  let priceDisclaimerPosition = 'top';
  let hasPartNumber = false;
  let hasPriceLabel = false;
  let hasComingSoon = false;
  let tabTitleFontFamily = 'bold';
  let accessoryCardBg = isDark ? 'black' : 'default';
  let accessoryCardWidth;
  let showOverlay = true;
  let showDetailsCard = false;
  let sizing;
  let hasImageNotFound = false;
  let disclaimerMargins;

  const comingSoonLabel = t('Shared.Common.comingSoonLabel');
  let priceDisclaimerLabel = t('Pages.Models.Exploration.msrpLabel');
  let buttonLabel = t('Shared.Common.submitButton');
  const partNumberLabel = t('Pages.Models.Exploration.partNumberLabel');
  const unableToLoadLabel = t('Pages.Models.Exploration.unableToLoadLabel');
  const pspMsrpLabel = t('Pages.Models.Exploration.msrpLabel');
  let dropdownDescription = '';
  let dropdownDescriptionStyles = {};
  let descriptionCopyOverflowStyles = {};
  let gtmInteractionType;
  let imageNotFoundLabel = '';

  switch (variant) {
    // PSP_SITE_NAME
    case variant2:
      accessoryFormatter = formatAccessoriesGenericCategory;
      transmissionStateKey = 'submittedTransmission';
      trimPathKey = 'detIdentifier';
      trimTransmissionPathKey = 'defaultTransmission.detIdentifier';
      optionTrimValuePath = 'detIdentifier';
      accessoryCardBg = 'white';
      showTabs = false;
      showAccordion = false;
      showSlider = true;
      dropdownKeys = [TRIM_DROPDOWN_KEY, TRANSMISSION_DROPDOWN_KEY, ACCESSORY_TYPE_DROPDOWN_KEY];
      defaultTab = ALL_TAB;
      defaultTabs = [];
      priceDisclaimerLabel = t('Pages.Models.Exploration.accessoriesDisclaimerText');

      buttonLabel = t('Pages.Models.Exploration.viewAccessoriesButton');
      dropdownList.splice(
        0,
        dropdownList.length,
        ...[
          {
            key: TRIM_DROPDOWN_KEY,
            label: t('Pages.Models.Exploration.accessoriesModelLabel'),
            placeholder: t('Pages.Models.Exploration.accessoriesModelLabel'),
          },
          {
            key: TRANSMISSION_DROPDOWN_KEY,
            label: t('Pages.Models.Exploration.accessoriesTransmissionLabel'),
            placeholder: t('Pages.Models.Exploration.accessoriesTransmissionLabel'),
          },
          {
            key: ACCESSORY_TYPE_DROPDOWN_KEY,
            label: t('Pages.Models.Exploration.accessoriesTypeLabel'),
            placeholder: '',
          },
        ],
      );
      hasAsterisk = true;
      priceDisclaimerPosition = 'top';
      disclaimerMargins = ['24px', '24px', '48px'];
      hasPartNumber = true;
      hasPriceLabel = true;
      hasComingSoon = true;
      hasImageNotFound = true;
      accessoryCardWidth = '280px';
      dropdownDescription = t('Pages.Models.Exploration.accessoriesBlurbText');
      imageNotFoundLabel = t('Pages.Models.Exploration.imageNotAvailableLabel');
      dropdownDescriptionStyles = { maxWidth: '1000px' };
      descriptionCopyOverflowStyles = {
        overflow: 'hidden',
        display: '-webkit-box',
        '-webkit-line-clamp': '3',
        '-webkit-box-orient': 'vertical',
      };
      showOverlay = false;
      showDetailsCard = true;
      sizing = 'compactDescription';
      gtmInteractionType = 'cta: click';
      break;

    // ACURA_SITE_NAME
    case variant1:
      tabTitleFontFamily = 'heading';
      accessoryCardBg = 'white';
      gtmInteractionType = 'accessories - accessory filter';
      break;

    // HONDA_SITE_NAME
    case variant3:
      gtmInteractionType = 'accessories - accessory filter';
      break;
    default:
      break;
  }

  return {
    accessoryFormatter,
    defaultTab,
    defaultTabs,
    descriptionCopyOverflowStyles,
    sizing,
    disclaimerMargins,
    dropdownDescription,
    dropdownDescriptionStyles,
    dropdownKeys,
    dropdownList,
    hasAsterisk,
    hasComingSoon,
    hasImageNotFound,
    hasPartNumber,
    hasPriceLabel,
    isDark,
    optionTrimValuePath,
    priceDisclaimerPosition,
    showAccordion,
    showDetailsCard,
    showOverlay,
    showSlider,
    showTabs,
    transmissionStateKey,
    trimPathKey,
    trimTransmissionPathKey,
    dictionary: {
      buttonLabel,
      comingSoonLabel,
      partNumberLabel,
      priceDisclaimerLabel,
      pspMsrpLabel,
      unableToLoadLabel,
      imageNotFoundLabel,
    },
    styles: {
      accessoryCardBg,
      accessoryCardWidth,
      tabTitleFontFamily,
    },
    gtmInteractionType,
  };
};

const getTrims = models => {
  if (!models) return undefined;
  return flatten(models?.map(model => flatten(model?.trims)));
};

/**
 * Formatter which maps fields for PSP due to structural differences with Autos
 *
 * @param {array} fields - The object containing Accessories component fields
 *  @returns {object} - Returns an updated fields object for use on PSP
 */
/* eslint-disable no-param-reassign */
export const mapPSPFields = fields => {
  if (!fields || isEmpty(fields) || !fields.modelYear) {
    return {};
  }

  fields.modelYear = {
    ...fields?.modelYear,
    fields: {
      ...fields?.modelYear?.fields,
      defaultTrim: { fields: fields?.modelYear?.fields?.models?.[0]?.trims?.[0] },
      trims: getTrims(fields?.modelYear?.fields?.models),
      model: {
        detKey: { value: fields?.modelYear?.fields?.models?.[0]?.detKey?.value },
      },
    },
  };

  return fields;
};
