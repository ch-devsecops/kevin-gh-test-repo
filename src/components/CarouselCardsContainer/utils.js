export const cardsPerPageParam = {
  one: 1,
  two: 2,
  three: 3,
  four: 4,
};

export const getCarouselPagesNumber = (itemsCount, itemsPerPage) => {
  const quotient = Math.floor(itemsCount / itemsPerPage);
  const remainder = itemsCount % itemsPerPage;
  return quotient + (remainder > 0 ? 1 : 0);
};

export const getResponsiveCardsPerPage = (itemsCount, cardsPerPage, isMobile, isTablet) => {
  let cards = cardsPerPage;
  cards = isMobile ? 1 : cards;
  cards = isTablet && cards > 3 ? 3 : cards;

  return itemsCount < cards ? itemsCount : cards;
};

export default { getCarouselPagesNumber, getResponsiveCardsPerPage };
