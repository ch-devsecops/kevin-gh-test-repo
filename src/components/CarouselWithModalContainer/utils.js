/**
 * returns the responsive number of cards per page
 * @param {number} itemsCount count of items
 * @param {number} cardsPerPage number of cards to show per page based on sitecore params
 * @param {boolean} isMobile true for mobile
 * @returns number of cards per page based on viewport or device
 */
const getResponsiveCardsPerPage = (itemsCount, cardsPerPage, isMobile) => {
  const cards = isMobile ? 1 : cardsPerPage;

  return itemsCount < cards ? itemsCount : cards;
};

export default getResponsiveCardsPerPage;
