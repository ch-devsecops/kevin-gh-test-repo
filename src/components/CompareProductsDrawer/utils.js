import { flattenDeep } from 'lodash/array';
import getValueByKey from '../../utils/getValueByKey';

/**
 * returns the miniature product card data for selected products to compare
 * @param detIds
 * @param data
 * @returns {Array}
 */
function getModels(detIds, data) {
  return flattenDeep(
    data?.map(category =>
      category?.subCategories?.map(subCategory =>
        subCategory?.models?.filter(model => detIds.includes(model?.detIdentifier?.value)),
      ),
    ),
  );
}

/**
 * maps marine web catalog to the expected data structure by the UI component
 * @param outboards
 * @returns {*}
 */
export function getMarineMappedCatalogData(outboards) {
  return outboards?.map(item => ({
    subCategories: [
      {
        id: item?.id,
        models: item?.models?.map(model => ({
          ...model,
          thumbnail: {
            ...model?.primaryImage?.item?.value,
            url: model?.primaryImage?.item?.value.src,
            alt: { value: model?.primaryImage?.item?.value.alt },
          },
        })),
      },
    ],
  }));
}

/**
 * maps engine web catalog to the expected data structure by the UI component
 * @param series
 * @returns {Array}
 */
export function getEngineMappedCatalogData(series) {
  return flattenDeep(
    series?.map(s =>
      s.crankshafts?.map(crankshaft => ({
        subCategories: [
          {
            id: crankshaft?.id,
            models: crankshaft?.models?.map(model => ({
              ...model,
              thumbnail: getValueByKey(model, 'images')?.[0],
            })),
          },
        ],
      })),
    ),
  );
}

export default getModels;
