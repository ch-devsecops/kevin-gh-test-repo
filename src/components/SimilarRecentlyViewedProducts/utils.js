/**
 * flatten the nest properties of engine models
 * @param {array<string>}detIds
 * @param {array<string>}data
 * @returns {*[]}
 */
export function getEngineModels(detIds, data) {
  const ret = [];
  detIds?.forEach(detId => {
    data?.forEach(category => {
      category?.crankshafts?.forEach(crankshaft => {
        const match = crankshaft?.models?.find(
          model => model?.detIdentifier?.value === detId && !ret?.find(item => item?.detIdentifier?.value === detId),
        );

        if (match) {
          match.seriesName = category.name;
          match.crankshaftName = crankshaft.name;
          ret.push(match);
        }
      });
    });
  });
  return ret;
}

/**
 * flatten the nest properties of marine models
 * @param detIds
 * @param data
 * @returns {*[]}
 */
export function getMarineModels(detIds, data) {
  const ret = [];
  detIds?.forEach(detId => {
    data?.forEach(category => {
      const match = category?.models?.find(
        model => model?.detIdentifier?.value === detId && !ret?.find(item => item?.detIdentifier?.value === detId),
      );

      if (match) {
        match.seriesName = category.name;
        match.crankshaftName = category.models.name;
        ret.push(match);
      }
    });
  });
  return ret;
}
