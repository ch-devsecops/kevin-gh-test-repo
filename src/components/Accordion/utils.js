import { ENGINE_SITE_NAME, MARINE_SITE_NAME } from '../../utils/constants';
import { getGtmTagValue } from '../../utils/gtmEvents';

/**
 * Creates GTM attributes for a productName
 * @param {string} productName
 * @param {object} gtmTags
 * @returns gtmAttributes
 */
export const getAccordionGtmTagsByProductName = ({ appName, gtmTags }) => {
  switch (appName) {
    case MARINE_SITE_NAME:
      return Object.freeze({
        'data-gtm-interaction-type': 'primary model nav',
        'data-gtm-component-type': getGtmTagValue(gtmTags?.componentName),
        'data-gtm-title': getGtmTagValue(gtmTags?.gtmTitle),
        'data-gtm-accordion-level': 'L1',
      });
    case ENGINE_SITE_NAME:
      return Object.freeze({
        'data-gtm-interaction-type': getGtmTagValue(gtmTags?.interactionType),
        'data-gtm-title': getGtmTagValue(gtmTags?.gtmTitle),
        'data-gtm-component-type': getGtmTagValue(gtmTags?.componentName),
        'data-gtm-body-style': getGtmTagValue(gtmTags?.bodyStyle),
        'data-gtm-model': getGtmTagValue(gtmTags?.modelName),
        'data-gtm-trim': getGtmTagValue(gtmTags?.trimName),
      });
    default:
      return {};
  }
};

export const getGtmTags = (category, componentName) =>
  Object.freeze({
    'data-gtm-category': getGtmTagValue(category),
    'data-gtm-component-type': getGtmTagValue(componentName),
  });
