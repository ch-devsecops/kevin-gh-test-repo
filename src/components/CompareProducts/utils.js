import { Node, Tree } from '@lukeaus/plain-tree';
import getValueByKey from '../../utils/getValueByKey';
import { urlProductParamNameMap } from '../../utils/constants';

/**
 * Map engine web catalog to the expected data structure by UI component.
 * @param {Object[]} series - The series with all the crankshafts and models.
 * @returns {Tree} engine catalog data in tree structure
 */
export const getEngineMappedCatalogData = series => {
  if (!Array.isArray(series) || !series?.length) return undefined;
  const rootNode = new Node('root');
  const tree = new Tree(rootNode);
  series?.forEach(item => {
    const seriesNode = rootNode.addChild({
      value: item?.name,
      text: item?.name,
      id: item?.id,
    });

    item?.crankshafts?.forEach(crankshaft =>
      crankshaft?.models?.forEach(model => {
        seriesNode.addChild({
          text: model?.name,
          value: model?.detIdentifier?.value,
          name: model?.name,
          detIdentifier: model?.detIdentifier?.value,
          model: crankshaft?.name,
          modelPage: model?.modelPage?.fields,
          thumbnail: getValueByKey(model, 'images')?.[0],
          id: model?.id,
        });
      }),
    );
  });
  return tree;
};

/**
 * Map marine web catalog to the expected data structure by UI component.
 * @param {Object[]}outboards
 * @returns {Tree} marine catalog data in tree structure
 */
export const getMarineMappedCatalogData = outboards => {
  if (!Array.isArray(outboards) || !outboards?.length) return undefined;
  const rootNode = new Node('root');
  const tree = new Tree(rootNode);
  outboards?.forEach(outboard => {
    const outboardsNode = rootNode.addChild({
      value: outboard?.name,
      text: outboard?.name,
      id: outboard?.id,
    });

    outboard?.models?.forEach(model => {
      const thumbnail = getValueByKey(model, 'primaryImage')?.[0]?.item?.value;
      outboardsNode.addChild({
        text: model?.name,
        value: model?.detIdentifier?.value,
        name: model?.modelName?.value,
        detIdentifier: model?.detIdentifier?.value,
        model: model?.name,
        modelPage: model?.modelPage?.fields,
        thumbnail: {
          ...thumbnail,
          url: thumbnail?.src,
        },
        id: model?.id,
      });
    });
  });
  return tree;
};

/**
 * Map PSP web catalog to the expected data structure by UI component.
 * @param {Object[]} categories
 * @returns { Tree } psp catalog data in tree structure
 */
export const getPspMappedCatalogData = vehicleTypesWithCategories => {
  if (!Array.isArray(vehicleTypesWithCategories) || !vehicleTypesWithCategories?.length) return undefined;
  const rootNode = new Node('root');
  const tree = new Tree(rootNode);

  vehicleTypesWithCategories.forEach(v => {
    const vehicleType = v?.name?.toUpperCase();
    const categories = v?.categories;
    categories?.forEach(category => {
      const categoryNode = rootNode.addChild({ text: category?.name, value: category?.categoryName?.value });
      category?.modelYears?.forEach(modelYear => {
        const yearNode = categoryNode.addChild({ text: modelYear?.year?.value, value: modelYear?.year?.value });
        modelYear?.models?.forEach(model =>
          model?.trims?.forEach(trim => {
            const trimNode = yearNode.addChild({
              text: trim?.name,
              value: trim?.detIdentifier?.value,
              name: trim?.name,
              detIdentifier: trim?.detIdentifier?.value,
              detKey: trim?.detKey?.value,
              modelPage: modelYear?.modelYearPage?.fields,
              thumbnail: trim?.primaryThumbnail?.item?.value,
              id: trim?.sotId?.value,
              modelKey: model?.detKey?.value,
              modelYear: modelYear?.year?.value,
              vehicleType,
            });
            trim?.transmissions?.[0]?.items.forEach(transmission => {
              trimNode.addChild({
                text: transmission?.name,
                value: transmission?.transmissionName?.value,
                modelCode: transmission?.modelCode,
                detIdentifier: transmission?.detIdentifier?.value,
                defaultExteriorColor: transmission?.defaultExteriorColor?.fields,
                hidePriceAndFinance: transmission?.hidePriceAndFinance?.value,
              });
            });
          }),
        );
      });
    });
  });

  return tree;
};

export const getUrlParamName = appName => urlProductParamNameMap.get(appName) || '';
