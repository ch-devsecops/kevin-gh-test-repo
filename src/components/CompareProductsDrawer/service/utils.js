import { flattenDeep } from 'lodash/array';
import getValueByKey from '../../../utils/getValueByKey';
import { MC_PRODUCT_NAME } from '../../../utils/constants';

/**
 * returns the miniature product card data for selected products to compare
 * @param detIds - model DET IDs
 * @param data - product catalog data
 * @returns {Array}
 */
export function getEngineModels(detIds, data) {
  return flattenDeep(
    data?.map(category =>
      category?.subCategories?.map(subCategory =>
        subCategory?.models?.filter(model => detIds.includes(model?.detIdentifier?.value)),
      ),
    ),
  );
}

/**
 * returns the miniature product card data for selected products to compare
 * @param detIds - transmission DET IDs
 * @param data - mapped product catalog data, from getPspMappedCatalogData
 * @returns {Array}
 */
export function getPspModels(detIds, data) {
  const models = flattenDeep(
    data?.map(category =>
      category?.subCategories?.map(subCategory =>
        subCategory?.models?.filter(model => model.transmissions?.some(item => detIds?.includes(item))),
      ),
    ),
  );
  // ensures models are returned in the same order as the provided detIds
  return detIds.map(detId => models.find(model => model.transmissions.includes(detId))).filter(id => !!id);
}

/**
 * maps marine web catalog to the expected data structure by the UI component
 * @param outboards
 * @returns {*}
 */
export function getMarineMappedCatalogData(outboards) {
  return outboards?.map(item => ({
    subCategories: [
      {
        id: item?.id,
        models: item?.models?.map(model => ({
          ...model,
          thumbnail: {
            ...model?.primaryImage?.item?.value,
            url: model?.primaryImage?.item?.value.src,
            alt: { value: model?.primaryImage?.item?.value.alt },
          },
        })),
      },
    ],
  }));
}

/**
 * maps engine web catalog to the expected data structure by the UI component
 * @param series
 * @returns {Array}
 */
export function getEngineMappedCatalogData(series) {
  return flattenDeep(
    series?.map(s =>
      s.crankshafts?.map(crankshaft => ({
        subCategories: [
          {
            id: crankshaft?.id,
            models: crankshaft?.models?.map(model => ({
              ...model,
              thumbnail: getValueByKey(model, 'images')?.[0],
            })),
          },
        ],
      })),
    ),
  );
}

/**
 * Returns a new array of vehicles with their associated product line.
 *
 * @function
 * @param {Object} vehicleData - The source data containing product lines and their associated vehicle types.
 * @param {string} vehicleData[].name - The name of the product line.
 * @param {Object[]} vehicleData[].vehicleTypes - The types of vehicles under the product line.
 *
 * @returns {Object[]} - An array of vehicles with added 'productLine' property.
 */
const getVehiclesWithProductLine = vehicleData => {
  if (!vehicleData || typeof vehicleData !== 'object') return [];

  let result = [];
  Object.values(vehicleData).forEach(({ name: productLine, vehicleTypes }) => {
    if (Array.isArray(vehicleTypes)) {
      const vehicles = vehicleTypes.map(vehicleType => ({ ...vehicleType, productLine }));
      result = result.concat(vehicles);
    }
  });

  return result;
};
/**
 * maps psp web catalog to the expected data structure by the UI component
 * @param categories
 * @returns {Array}
 */
export const getPspMappedCatalogData = fieldsData => {
  const vehicles = getVehiclesWithProductLine(fieldsData);
  return flattenDeep(
    vehicles?.map(vehicle => {
      const productline = vehicle?.productLine === MC_PRODUCT_NAME ? 'Mc' : 'AtvSxs';
      return vehicle?.categories?.map(category =>
        category.modelYears?.map(modelYear =>
          modelYear?.models?.map(model => ({
            subCategories: [
              {
                id: model?.sotId,
                models: model?.trims
                  ? model.trims.map(trim => ({
                      ...trim,
                      modelName: {
                        value: trim?.trimName?.value,
                      },
                      thumbnail: {
                        ...trim?.primaryThumbnail?.item?.value?.images?.[0],
                      },
                      productline,
                      // store all transmission IDs under this trim/model
                      transmissions: trim?.transmissions?.[0]?.items?.map(tran => tran?.detIdentifier?.value),
                    }))
                  : [],
              },
            ],
          })),
        ),
      );
    }),
  );
};

/**
 * Generates a URL for the compare page based on the provided data.
 *
 * @param {Array} data - An array of objects containing product data.
 * @returns {string} The URL for the compare page, or an empty string if the data is invalid.
 */
export const getComparePageUrl = data => {
  if (data.length === 0) {
    return '';
  }
  const firstProductline = data[0].productline;
  const isProductlineSame = data.every(item => item.productline === firstProductline);

  if (!isProductlineSame) return '';

  return `Shared.Common.comparePage${firstProductline}Url`;
};
