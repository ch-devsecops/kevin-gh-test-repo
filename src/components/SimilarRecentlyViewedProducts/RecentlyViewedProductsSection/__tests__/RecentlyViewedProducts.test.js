import React from 'react';
import { axe, cleanup, jssRender, screen } from 'test-utils';
import RecentlyViewedProducts from '..';
import LocalStorageContext from '../../../LocalStorageContext/LocalStorageContext';
import getValueByKey from '../../../../utils/getValueByKey';
import { getEngineModels } from '../../utils';

afterEach(cleanup);

const mockData = require('../../../../../sample-mock-data/routes/recently-viewed/en.json');

const models = getValueByKey(mockData, 'series');

const localStorageStateMock = {
  recentlyViewedProducts: ['12101', '12099'],
  setRecentlyViewedProducts: () => {},
  toCompareProducts: [],
  setToCompareProducts: () => {},
  removeFromCompareProducts: () => {},
};

describe('<RecentlyViewedProducts>', () => {
  test('should render correctly', () => {
    jssRender(
      <LocalStorageContext.Provider value={localStorageStateMock}>
        <RecentlyViewedProducts models={models} getModels={getEngineModels} />
      </LocalStorageContext.Provider>,
    );

    expect(screen.queryByText('GX340')).toBeInTheDocument();
  });

  test('should not render when there is no item in the localstorage', () => {
    jssRender(
      // eslint-disable-next-line react/jsx-no-constructed-context-values
      <LocalStorageContext.Provider value={{ ...localStorageStateMock, recentlyViewedProducts: [] }}>
        <RecentlyViewedProducts models={models} getModels={getEngineModels} />
      </LocalStorageContext.Provider>,
    );

    expect(screen.queryByText('GX340')).not.toBeInTheDocument();
  });

  test('should return null if there is no fields', () => {
    const { container } = jssRender(<RecentlyViewedProducts models={null} getModels={getEngineModels} />);

    expect(container).toMatchSnapshot();
  });

  test('should have no accessibility violations', async () => {
    const { container } = jssRender(
      <LocalStorageContext.Provider value={localStorageStateMock}>
        <RecentlyViewedProducts models={models} getModels={getEngineModels} />
      </LocalStorageContext.Provider>,
    );

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  });
});
