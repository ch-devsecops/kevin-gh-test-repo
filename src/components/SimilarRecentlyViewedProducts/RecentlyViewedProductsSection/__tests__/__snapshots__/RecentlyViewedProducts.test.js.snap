// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecentlyViewedProducts> should have no accessibility violations 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 hwpdqT"
    display="flex"
    width="1,1,0.5"
  >
    <h4
      aria-label="Pages.Models.Exploration.recentlyViewedProductsLabel"
      class="sc-1k44dfv-0 chqxmS"
    >
      Pages.Models.Exploration.recentlyViewedProductsLabel
    </h4>
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 heIgvm"
    >
      <div
        class="Box-sc-1bsf784-0 sc-1k44dfv-0 fZLgxb"
      >
        <div
          class="Box-sc-1bsf784-0 sc-1k44dfv-0 ljGqou"
          data-testid="product-card"
        >
          <div
            style="overflow: hidden; transition-property: height; display: block; transition-duration: 0.25s;"
          />
          <div
            style="overflow: hidden; transition-property: height; display: block; transition-duration: 0.25s;"
          >
            <a
              class="Link-sc-16wd4a0-0 bXBGRT"
              data-gtm-body-style="GX Series"
              data-gtm-interaction-type="explore"
              data-gtm-model="VerticalCrankshaft"
              data-gtm-trim="GXV160"
              title="view details"
            >
              <div
                class="sc-1k44dfv-0 cALsOM"
                mb="xs,s"
                mt="0"
              >
                

                <h5
                  class="sc-1k44dfv-0 gXWcpP"
                >
                  GXV160
                </h5>
              </div>
            </a>
          </div>
          <a
            class="Link-sc-16wd4a0-0 bXBGRT"
            data-gtm-body-style="GX Series"
            data-gtm-interaction-type="explore"
            data-gtm-model="VerticalCrankshaft"
            data-gtm-trim="GXV160"
            title="view details"
          >
            <div
              class="Box-sc-1bsf784-0 jgcAXQ"
              height="139px,248px"
              order="-1"
            >
              <img
                alt="Front of Honda GXV160 engine"
                class="sc-1k44dfv-0 jLTFZk"
                draggable="true"
                loading="lazy"
                src="https://engine-dit-01-dmm.engine.honda.ca/-/media/Brands/Engine/21-GX-Series-Listing/GXV-160/GXV160.jpg?rev=4870392a1a0148f3bfcb6a0f9ab0514b"
                style="object-fit: contain;"
              />
            </div>
          </a>
          <div
            class="Box-sc-1bsf784-0 QoaNQ"
          >
            <div
              class="sc-1k44dfv-0 ktMwmo"
            >
              

              <p>
                Solid performance, reliability and fuel efficiency. Ideal for lawn, construction and agriculture.
              </p>
            </div>
          </div>
          <a
            class="Button-pulw6t-0 sc-1k44dfv-0 eIRfAq"
            data-gtm-body-style="GX Series"
            data-gtm-interaction-type="explore"
            data-gtm-model="VerticalCrankshaft"
            data-gtm-title="view details"
            data-gtm-trim="GXV160"
            data-testid="dsr-button-component"
            display="none,block"
            tabindex="0"
          >
            Shared.Common.viewDetailsButton
          </a>
          <div
            class="Box-sc-1bsf784-0 sc-1k44dfv-0 jkGzMS"
          >
            <label
              class="sc-bwzfXH cvvuMq"
              data-gtm-body-style="GX Series"
              data-gtm-interaction-type="compare item click"
              data-gtm-model="VerticalCrankshaft"
              data-gtm-title="added comparison item"
              data-gtm-trim="GXV160"
              data-testid="product-card-compare-btn"
              data-testvalue="false"
            >
              <input
                aria-checked="false"
                aria-label="Shared.Common.compareCheckboxAria"
                autocomplete="off"
                class="sc-bdVaJa gBQUrE"
                data-gtm-body-style="GX Series"
                data-gtm-interaction-type="compare item click"
                data-gtm-model="VerticalCrankshaft"
                data-gtm-title="added comparison item"
                data-gtm-trim="GXV160"
                id="9517794341634F308D3315497A75A138"
                tabindex="-1"
                type="checkbox"
              />
              <span
                class="sc-htpNat hVaxDi"
                data-testid="cy-compare-toggle"
              />
              <label
                class="Label-sc-115wxjh-0 sc-1k44dfv-0 jLfsOv"
                for="9517794341634F308D3315497A75A138"
              >
                Shared.Common.addToCompareButton
              </label>
            </label>
          </div>
        </div>
      </div>
      <div
        class="Box-sc-1bsf784-0 sc-1k44dfv-0 fZLgxb"
      >
        <div
          class="Box-sc-1bsf784-0 sc-1k44dfv-0 ljGqou"
          data-testid="product-card"
        >
          <div
            style="overflow: hidden; transition-property: height; display: block; transition-duration: 0.25s;"
          />
          <div
            style="overflow: hidden; transition-property: height; display: block; transition-duration: 0.25s;"
          >
            <a
              class="Link-sc-16wd4a0-0 bXBGRT"
              data-gtm-body-style="GX Series"
              data-gtm-interaction-type="explore"
              data-gtm-model="HorizontalCrankshaft"
              data-gtm-trim="GX340"
              title="view details"
            >
              <div
                class="sc-1k44dfv-0 cALsOM"
                mb="xs,s"
                mt="0"
              >
                

                <h5
                  class="sc-1k44dfv-0 gXWcpP"
                >
                  GX340
                </h5>
              </div>
            </a>
          </div>
          <a
            class="Link-sc-16wd4a0-0 bXBGRT"
            data-gtm-body-style="GX Series"
            data-gtm-interaction-type="explore"
            data-gtm-model="HorizontalCrankshaft"
            data-gtm-trim="GX340"
            title="view details"
          >
            <div
              class="Box-sc-1bsf784-0 jgcAXQ"
              height="139px,248px"
              order="-1"
            >
              <img
                alt="Front of Honda GX340 engine"
                class="sc-1k44dfv-0 jLTFZk"
                draggable="true"
                loading="lazy"
                src="https://engine-dit-01-dmm.engine.honda.ca/-/media/Brands/Engine/21-GX-Series-Listing/GX-340/GX340.jpg?rev=f3220fba10844848aa306d90418adfcf"
                style="object-fit: contain;"
              />
            </div>
          </a>
          <div
            class="Box-sc-1bsf784-0 QoaNQ"
          >
            <span
              class="sc-1k44dfv-0 ktMwmo"
            >
              

            </span>
          </div>
          <a
            class="Button-pulw6t-0 sc-1k44dfv-0 eIRfAq"
            data-gtm-body-style="GX Series"
            data-gtm-interaction-type="explore"
            data-gtm-model="HorizontalCrankshaft"
            data-gtm-title="view details"
            data-gtm-trim="GX340"
            data-testid="dsr-button-component"
            display="none,block"
            tabindex="0"
          >
            Shared.Common.viewDetailsButton
          </a>
          <div
            class="Box-sc-1bsf784-0 sc-1k44dfv-0 jkGzMS"
          >
            <label
              class="sc-bwzfXH cvvuMq"
              data-gtm-body-style="GX Series"
              data-gtm-interaction-type="compare item click"
              data-gtm-model="HorizontalCrankshaft"
              data-gtm-title="added comparison item"
              data-gtm-trim="GX340"
              data-testid="product-card-compare-btn"
              data-testvalue="false"
            >
              <input
                aria-checked="false"
                aria-label="Shared.Common.compareCheckboxAria"
                autocomplete="off"
                class="sc-bdVaJa gBQUrE"
                data-gtm-body-style="GX Series"
                data-gtm-interaction-type="compare item click"
                data-gtm-model="HorizontalCrankshaft"
                data-gtm-title="added comparison item"
                data-gtm-trim="GX340"
                id="7B841133D8474EA0B701E13F8A9E8485"
                tabindex="-1"
                type="checkbox"
              />
              <span
                class="sc-htpNat hVaxDi"
                data-testid="cy-compare-toggle"
              />
              <label
                class="Label-sc-115wxjh-0 sc-1k44dfv-0 jLfsOv"
                for="7B841133D8474EA0B701E13F8A9E8485"
              >
                Shared.Common.addToCompareButton
              </label>
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<RecentlyViewedProducts> should return null if there is no fields 1`] = `<div />`;
